light-zk.js/src/transaction/indexRecentTransactions.ts

```diff
@@ -308,58 +321,62 @@ const getTransactionsBatch = async ({
   );
 
   const lastSignature = signatures[signatures.length - 1];
-  let txs: (ParsedTransactionWithMeta | null)[] = [];
-  let index = 0;
   const signaturesPerRequest = 25;
 
-  while (index < signatures.length) {
-    try {
-      const txsBatch = await connection.getParsedTransactions(
-        signatures
-          .slice(index, index + signaturesPerRequest)
-          .map((sig) => sig.signature),
-        {
-          maxSupportedTransactionVersion: 0,
-          commitment: "confirmed",
-        },
-      );
-
-      if (!txsBatch.some((t) => !t)) {
-        txs = txs.concat(txsBatch);
-        index += signaturesPerRequest;
-      }
-    } catch (e) {
-      console.log("retry");
-      await sleep(2000);
-    }
-  }
+  let txs = [];
+  for (let i = 0; i < signatures.length; i += signaturesPerRequest) {
+    const sigsBatch = signatures
+      .slice(i, i + signaturesPerRequest)
+      .map((sig) => sig.signature);
 
-  const indexerEventTransactions = txs.filter((tx: any) => {
-    const accountKeys = tx.transaction.message.accountKeys;
-    const splNoopIndex = accountKeys.findIndex((item: ParsedMessageAccount) => {
-      const itemStr =
-        typeof item === "string" || item instanceof String
-          ? item
-          : item.pubkey.toBase58();
-      return itemStr === new PublicKey(SPL_NOOP_ADDRESS).toBase58();
-    });
-
-    if (splNoopIndex) {
-      return txs;
-    }
-  });
+    const txsBatch = await fetchTransactions(connection, sigsBatch);
+    txs.push(...txsBatch);
+  }
 
+  const filteredTxs = filterFetchedTransactions(txs);
   const indexerTransactionEvents = await processIndexerEventsTransactions(
-    indexerEventTransactions,
+    filteredTxs,
   );
 
   indexerTransactionEvents.forEach((event) => {
-    processIndexedTransaction(event!, transactions);
+    processIndexedTransaction(event, transactions);
   });
 
   return lastSignature;
 };
 
+const fetchTransactions = async (connection: Connection, sigs: string[]) => {
+  try {
+    const txsBatch = await connection.getParsedTransactions(sigs, {
+      maxSupportedTransactionVersion: 0,
+      commitment: "confirmed",
+    });
+
+    return txsBatch.every((t) => t) ? txsBatch : [];
+  } catch (e) {
+    console.log("retry");
+    await sleep(2000);
+    return [];
+  }
+};
+
+const filterFetchedTransactions = (txs: any[]) => {
+  return txs.filter((tx) => {
+    const accountKeys = tx.transaction.message.accountKeys;
+    const splNoopIndex = accountKeys.findIndex(
+      (item: { pubkey: { toBase58: () => any } }) => {
+        const itemStr =
+          typeof item === "string" || item instanceof String
```

Comment:
So here we can receive either a string or a key pair?

I'm fine with fixing that in an another PR, but of that's the case and since we are refactoring this part of the SDK, I think we should get rid of this error-prone duck typing and ensure we expect/accept just one type. WDYT?

Reply:
@vadorovsky
hard to deal with an answer. better show code:

```typescript
const filterFetchedTransactions = (txs: any[]): any[] => {
  return txs.filter((tx: any) => {
    const accountKeys = tx.transaction.message.accountKeys;
    const splNoopIndex = accountKeys.findIndex(
      (item: any) => {
        const itemStr = item.pubkey.toBase58();
        return itemStr === new PublicKey(SPL_NOOP_ADDRESS).toBase58();
      },
    );

    return splNoopIndex !== -1;
  });
};
```
In the updated code, I removed the duck-typing by removing the type annotation from the inner callback function. Instead of explicitly specifying the type of item as { pubkey: { toBase58: () => any } }, I could change it to `any`. This allows TypeScript to infer the type of item based on usage, avoiding the need for any explicit interface.

Please note that using any can bypass type checking, so it's recommended to use specific types whenever possible. However, if you need to avoid introducing new interfaces and the structure of the data is dynamic, using any can be a practical solution.